backend/src/hatchling/metadata/core.py

```diff
@@ -31,11 +33,101 @@
     import tomli as tomllib
 
 
+VARIANT_HASH_LEN = 8
+
+
 def load_toml(path: str) -> dict[str, Any]:
     with open(path, encoding='utf-8') as f:
         return tomllib.loads(f.read())
 
 
+@dataclass
+class VariantProviderConfig:
+    requires: list[str]
+    plugin_api: str | None = None
+    enable_if: str | None = None
+    optional: bool = False
+
+    @classmethod
+    def from_dict(cls, data: dict):
+        """Creates an instance of VariantProviderConfig from a dictionary."""
+        # Convert hyphenated keys to underscored keys
+        data = {key.replace("-", "_"): value for key, value in data.items()}
+
+        # Create an instance of VariantProviderConfig
+        return cls(**data)
+
+    def validate(self):
+        """Validates the VariantProviderConfig instance."""
+        if not self.requires:
+            raise ValueError("Requires list cannot be empty")
+
+
+@dataclass
+class VariantConfig:
+    vhash: str
+    properties: list[str]
+    default_priorities: dict[str, list[str]]
+    providers: dict[str, VariantProviderConfig]
+
+    @classmethod
+    def from_dict(cls, data: dict, vprops: list[str] | None,
+                  variant_label: str | None):
+        """Creates an instance of VariantConfig from a dictionary."""
+        data = data.copy()
+
+        if vprops is None:
+            data["vhash"] = None
+            data["properties"] = None
+
+
+        elif len(vprops) == 0:
+            data["vhash"] = "0" * VARIANT_HASH_LEN
+            data["properties"] = []
+
+        else:
+            # Normalizing
+            _vprops = [
+                [el.strip() for el in vprop.split("::")]
+                for vprop in vprops
+            ]
+            for vprop in _vprops:
+                assert len(vprop) == 3, f"Invalid variant property: {vprop}"
+
+            data["properties"] = [" :: ".join(vprop) for vprop in sorted(_vprops)]
+
+            hash_object = hashlib.sha256()
+            for vprop in data["properties"]:
+                hash_object.update(f"{vprop}\n".encode())
+            data["vhash"] = hash_object.hexdigest()[:VARIANT_HASH_LEN]
```

Comment:
Computing the hash seems pointless when `variant_label is not None` and it is overwritten afterwards.